doc/man3/SSL_shutdown.pod

```diff
@@ -107,6 +107,8 @@ and the peer's "close notify" alert was received.
 The shutdown was not successful because a fatal error occurred either
 at the protocol level or a connection failure occurred. It can also occur if
 action is need to continue the operation for non-blocking BIOs.
+It can also happen in a bidirectional shutdown when the peer sends more data
+instead of the expected "close notify" shutdown alert.
 Call L<SSL_get_error(3)> with the return value B<ret>
 to find out the reason.
```

Comment:
I think the "fatal error" above is misleading. Or I don't understand what it's trying to say. In any case I think this can be reworded better.

Reply:
This paragraph is about when -1 is returned from SSL_shutdown:
I think a "fatal error" might mean a malformed packet for instance?
A connection failure is termination of the socket due to a RST packet for instance?
But a SSL_ERROR_SYSCALL kind of error can as well happen if a
data packet arrives after the close notify alert is sent.
That happens when there is a race condition between SSL_shutdown
vs. SSL_write.
Do you have an idea how to put the words better?

Reply:
To me "fatal error" means it's never going to return 1, and that for whatever reason we can't continue with the shutdown. But it seems to me that it might return -1 for various reasons, and that like always you should just check SSL_get_error() to see if it's a fatal error or not. But it's unclear how to know that you need to call SSL_read() instead before SSL_shutdown() can return 1.

Reply:
Or should you call SSL_read() until that returns -1 and SSL_get_error() returns SSL_ERROR_ZERO_RETURN?

Reply:
Actually I never thought of the possibility to call SSL_read after the first SSL_shutdown.
to avoid this race condition.
This could work.

Reply:
So if there is pending data the second SSL_shutdown will return SSL_ERROR_SYSCALL
which is fatal.  But that can be avoided by reading all remaining data until
SSL_read returns SSL_ERROR_ZERO_RETURN.  Sounds good.

Reply:
If SSL_read returns SSL_ERROR_ZERO_RETURN then the second SSL_shutdown
will succeed immediately, and is probably not even necessary?

Reply:
a quick experiment shows it is possible to send data after SSL_read returns ZERO_RETURN
and to receive the data on the other side, avoiding the -1 from SSL_shutdown

Reply:
actually the return value from SSL_read is 0 and SSL_get_error() returns SSL_ERROR_ZERO_RETURN.

Reply:
Documentation updated.